pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,728 @@
+func (i *GroupEntityIndex) syncSelector(sItem *selectorItem) bool {
+	updated := false
+
+	scanLabels := func(labelItemKeys sets.String) {
+		for lKey := range labelItemKeys {
+			lItem := i.labelItems[lKey]
+			if i.match(lItem.entityType, lItem.labels, lItem.namespace, sItem.selector) {
+				// Connect the selector and the label if they didn't match before, otherwise do nothing.
+				if !sItem.labelItemKeys.Has(lKey) {
+					sItem.labelItemKeys.Insert(lKey)
+					lItem.selectorItemKeys.Insert(sItem.selector.NormalizedName)
+					updated = true
+				}
+			} else {
+				// Disconnect the selector and the label if they matched before, otherwise do nothing.
+				if sItem.labelItemKeys.Has(lKey) {
+					sItem.labelItemKeys.Delete(lKey)
+					lItem.selectorItemKeys.Delete(sItem.selector.NormalizedName)
+					updated = true
+				}
+			}
+		}
+	}
+
+	// By default, the selector selects Pods. It selects ExternalEntities only if ExternalEntitySelector is set
+	// explicitly.
+	entityType := podEntityType
+	if sItem.selector.ExternalEntitySelector != nil {
+		entityType = externalEntityType
+	}
+	// If the selector is Namespace scoped, it can only match labelItems in the Namespace. Otherwise, all labelItems
+	// should be checked.
+	if sItem.selector.Namespace != "" {
+		labelItemKeys, _ := i.labelItemIndex[entityType][sItem.selector.Namespace]
+		scanLabels(labelItemKeys)
+	} else {
+		for _, labelItemKeys := range i.labelItemIndex[entityType] {
+			scanLabels(labelItemKeys)
+		}
+	}
+	return updated
+}
+
+func (i *GroupEntityIndex) AddGroup(groupType string, name string, selector *types.GroupSelector) {
+	gKey := getGroupItemKey(groupType, name)
+	sKey := getSelectorItemKey(selector)
+
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	gItem, exists := i.groupItems[gKey]
+	if exists {
+		// Its selector doesn't change, do nothing.
+		if gItem.selectorItemKey == sKey {
+			return
+		}
+		i.deleteGroupFromSelectorItem(gItem.selectorItemKey, gKey)
+		gItem.selectorItemKey = sKey
+		gItem.selector = selector
+	} else {
+		gItem = &groupItem{
+			groupType:       groupType,
+			name:            name,
+			selector:        selector,
+			selectorItemKey: sKey,
+		}
+		i.groupItems[gKey] = gItem
+	}
+
+	// Create a selectorItem if it doesn't exist.
+	sItem, exists := i.selectorItems[sKey]
+	if !exists {
+		sItem = i.createSelectorItem(gItem)
+	}
+	sItem.groupItemKeys.Insert(gKey)
+}
+
+func (i *GroupEntityIndex) DeleteGroup(groupType string, name string) {
+	gKey := getGroupItemKey(groupType, name)
+
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	gItem, exists := i.groupItems[gKey]
+	if !exists {
+		return
+	}
+
+	// Delete the group from its associated selectorItem and groupItems.
+	i.deleteGroupFromSelectorItem(gItem.selectorItemKey, gKey)
+	delete(i.groupItems, gKey)
+}
+
+func (i *GroupEntityIndex) notify(selector string) {
+	sItem := i.selectorItems[selector]
+	for group := range sItem.groupItemKeys {
+		i.eventChan <- group
```

Comment:
This can be a block call? Does it make sense to call notify() out of locking?

Reply:
I used buffered channel to make it as quick as possible, but yes it can still block if it's not consumed fast enough and less efficient than just moving it out of locking anyway. Then maybe I could just make `notify` execute synchrounouly out of locking to make it simpler and easier to test.

Reply:
Just found we cannot remove locking here as it needs to access i.selectorItems and sItem.groupItemKeys.

Reply:
What you think about copying the keys and notify after releasing the lock? But if the channel does not really block, or non-blocking does not really help parallelization, might not be worthwhile.

Reply:
If we don't assume the channel can be full with 1000 buffers, it won't block, and copying the keys will introduce some overhead and may use more time than sending items to channels to directly. Maybe let's have a test to see if it can really block or we increate the buffer size later.

Reply:
Sounds good to me.

Reply:
Probably good to add some comments before notify() about the possible blocking.

Reply:
Done, and add comment to `AddEventHandler` that eventHandler shouldn't perform blocking operation. I could add time stats to log unexpected long executiont time later.